### PR TITLE
Fixed #181 -- Added copy-to-clipboard feature to code snippets

### LINKED ISSUE #181: Add "Copy" button to code snippets
Note: *This is more like a request than a issue*. 

It would be nice if some code snippets have a "Copy" button/link to quickly copy the code to clipboard.

Something like the image (I'm not that proficient in CSS BTW :smile:)

django

Great work and re-design!!

### PR DESCRIPTION
**Pull request by bmispelon**
The new clipboard feature is currently "opt-in" and documentation snippets that want to take advantage of it need to be modified like so:
```
..snippet
     :filename: path/to/file.py
     :clippify:
    ...
```
When clicked, the clipboard icon copies the text of the snippet to the clipboard and flashes a small "Copied!" message on the screen (see screenshots).
Since we use clipboardjs.com, the browser support should be pretty good (https://clipboardjs.com/#browser-support).

clippify1
clippify2

### CODE REVIEW

**Comment by bmispelon** [comment]
I added a workaround for safari where `executeCommand` doesn't quite work: it tells the user to press `⌘-C` to actually copy the code (this is how they handle it on the clipboardjs website).

**Comment by erikr** [comment]
Looks good to me! :shipit:

**Comment by timgraham** [comment]
Why did you decide for an opt-in approach as opposed to automatically applying this to all snippets? I'd think the latter would be simpler and easier in the long run.

**Comment by bmispelon** [comment]
I hesitated for a bit and decided it was better to opt-in so we can roll out the feature on a case-by-case basis.

I haven't written a lot of documentation so I'm not sure if that clipboard feature is desirable everywhere.

I'm happy to change the behavior to opt-out (via a `:noclippify:` directive), or even force it everywhere if you think that's a better approach.

**Comment by timgraham** [comment]
I don't think we'd need an option to disable it. If it makes it easier, this could apply to all highlighted blocks, not just snippets.

**Comment by bmispelon** [comment]
OK, I'll enable the feature on all code snippets.

Regarding enabling it to highlighted blocks too, I ran into two issues while looking into it:

1) It's not as obvious where to include the button because there's no "header" anymore
2) It doesn't make sense to copy/paste all blocks, and some blocks need to be partially copied. Take the example of the first block in [part 1 of the tutorial](https://docs.djangoproject.com/en/1.9/intro/tutorial01/):
```
$ python -m django --version
```
For this one, you don't want the user to copy the `$` sign at the beginning of the line.

**Comment by bmispelon** [comment]
I've updated the django/django#6875 PR to add the feature on all code snippets.

File: djangoproject/static/js/mod/clippify.js [hunk 1]
```diff
@@ -0,0 +1,22 @@
+define(['jquery', 'clipboard'], function($, Clipboard) {
+    $('.snippet').each(function(){
```
Comment (timgraham) [inline djangoproject/static/js/mod/clippify.js]: add space before { (a few cases of this)

File: djangoproject/static/js/mod/clippify.js [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+        var code = $('.highlight', this);
+        var btn = $('<span class="btn-clipboard" title="Copy this code">').append('<i class="icon icon-clipboard">').data('clipboard-text', code.text());
```
Comment (timgraham) [inline djangoproject/static/js/mod/clippify.js]: possible to multiline this? (having to scroll horizontally in review is annoying)

File: djangoproject/static/js/mod/clippify.js [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+        'text': function(trigger){
+            return $(trigger).data('clipboard-text');
+        }
+    });
+    clip.on('success', function(e){
+        var success = $('<span class="clipboard-success">').text('Copied!')
+        success.prependTo(e.trigger).delay(1000).fadeOut();
+    });
+    clip.on('error', function(e){
+        // Weird safari behavior
```
Comment (timgraham) [inline djangoproject/static/js/mod/clippify.js]: This comment could use a tad more explanation. (I guess this branch is for Safari only?)

**Comment by timgraham** [comment]
While testing this, I noticed that a blank newline appears when pasting a copied snippet. Is it possible to fix that?

**Comment by bmispelon** [comment]
@timgraham I've added a `$.trim()` to the copied text so it should strip any leading/trailing whitespace now (tested locally).

**Comment by timgraham** [comment]
LGTM

**Pull request merged**